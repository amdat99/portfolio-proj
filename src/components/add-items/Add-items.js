import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
// import FormInput from '../formInput/FormInput';
// import CustomButton from '../custom-button/CustomButton';
import { connect } from "react-redux";
import { setItemsDoc } from "../../firebase/firebase";
import { createStructuredSelector } from "reselect";
import {
  selectCurrentUser,
  selectProfileName,
} from "../../redux/user/user.selectors";
import { selectErrorMessage } from "../../redux/shop/shop.selectors";

import "./add-items.scss";

export const AddItems = ({ currentUser, error, profileName }) => {
  const [itemData, setItemData] = useState({
    name: "",
    price: "",
    description: "",
    category: "",
    soldBy: "",
    picture: "",
    userId: "",
    productId: "",
  });
  const [imageToggle, setImageToggle] = useState(false);
  const [submitToggle, setSubmitToggle] = useState(false);

  useEffect(() => {
    if (currentUser !== null) {
      setItemData({
        soldBy: profileName.toString(),
        userId: currentUser.profileId,
        productId: Math.random(),
      });
    }
  }, [currentUser, profileName]);

  const { name, price, description, picture } = itemData;

  const handleSubmit = async (event) => {
    event.preventDefault();
    // =>
    setItemsDoc(itemData);

    // if (error) {
    //   alert("there was an error", error.message);
    // } else {
    setItemData({
      name: "",
      price: "",
      description: "",
      category: "",
      picture: "",
      soldBy: profileName.toString(),
      userId: currentUser.profileId,
      productId: Math.random(),
    });
    setImageToggle(false);
    setSubmitToggle(false);
    alert("your listing was successfully updated");
  };

  const handleChange = (event) => {
    const { name, value } = event.target;
    setItemData({ ...itemData, [name]: value });
  };

  const verifyImage = () => {
    if (picture) {
      setItemData({ picture: "" });
      setImageToggle(false);
      setSubmitToggle(false);
      alert("your image is not valid");
    }
  };

  const toggleShowImage = () => {
    setImageToggle(true);

    const xhr = new XMLHttpRequest(); //verify image size
    xhr.open("GET", picture, true);
    xhr.responseType = "image/png";
    xhr.onreadystatechange = function () {
      if (this.readyState === this.DONE) {
        const size = this.response.byteLength;
        if (size > 2242880) {
          alert("use an image under 2mb");
        }
        setSubmitToggle(true);
      }
    };
    xhr.send(null);
  };

  const categories = ["electronics", "clothes", "shoes", "food", "general"];

  return (
    <div>
      <form className="sign-on-container" onSubmit={handleSubmit}>
        <h1 style={{ marginRight: "65px" }}>List your Item</h1>

        <input
          className="sign-on-input"
          type="text"
          name="name"
          placeholder="Enter the item name:"
          value={name}
          onChange={handleChange}
          label="name"
          required
        />

        <input
          className="sign-on-input"
          type="number"
          name="price"
          placeholder="Enter the price"
          value={price}
          onChange={handleChange}
          label="Price"
          required
        />

        <textarea
          className="sign-on-input add-items-description"
          type="text"
          name="description"
          placeholder="Add your description"
          value={description}
          onChange={handleChange}
          label="description"
          required
        />

        <p>Choose category</p>
        {categories.map((category, i) => (
          <div className="additems-categories" key={i}>
            {" "}
            {category}
            <input
              className=""
              type="radio"
              name="category"
              value={category}
              onChange={handleChange}
              label="category"
              required
            ></input>{" "}
          </div>
        ))}

        <input
          className="sign-on-input"
          type="url"
          name="picture"
          placeholder="enter a imageUrl"
          value={picture}
          onChange={handleChange}
          label="picture"
          required
        />
        <button type="button" onClick={toggleShowImage}>
          VerifyImage
        </button>

        {currentUser ? (
          <div>
            {submitToggle ? (
              <button className="add-item-submit" type="submit">
                ADD ITEM
              </button>
            ) : (
              <button
                type="button"
                onClick={() => alert("verify image to sumbit item")}
              >
                ADD ITEM
              </button>
            )}
          </div>
        ) : (
          <Link to="/signon"> Sign in to List Items</Link>
        )}
      </form>
      {imageToggle ? (
        <div>
          <img
            onError={verifyImage}
            src={picture}
            alt="item"
            width="220"
            height="240"
          />
        </div>
      ) : null}
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  error: selectErrorMessage,
  currentUser: selectCurrentUser,
  profileName: selectProfileName,
});

export default connect(mapStateToProps)(AddItems);
