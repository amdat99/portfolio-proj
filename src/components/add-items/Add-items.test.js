import React from 'react';

import { shallow,render,mount } from 'enzyme';
import {createSerializer} from 'enzyme-to-json';
import toJson from 'enzyme-to-json';

import {AddItems} from './Add-items'

expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));
// describe('cart item', () => {
      
        let mockhandleSubmit = jest.fn();
        let wrapper
        let wrapper2
    
        const mockOnChange = jest.fn();


        
  beforeEach(() => {

   const mockProps = {
            itemData:{ name: 'ball', price: 4, description: 'good', category: 'general', soldBy: 'john', picture: 'picyire', userId:'123',productId: 123},
            handleSubmit: mockhandleSubmit,
            submitToggle: true
            
          }; 
          
           wrapper = toJson(shallow(<AddItems  {...mockProps} />))
           wrapper2 = shallow(<AddItems {...mockProps} />)
        })

        afterEach(() => {
            jest.clearAllMocks();
        });
        
    it('expect to render additems component', () => {
    expect(wrapper).toMatchSnapshot();
    })

    // it('expect to render additems component', () => {

    //     wrapper2.find('.sign-on-container').simulate('submit');
    //     expect(mockhandleSubmit).toHaveBeenCalled()
    // })
