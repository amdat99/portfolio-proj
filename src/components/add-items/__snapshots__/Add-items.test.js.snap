// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render additems component 1`] = `
<div>
  <form
    className="sign-on-container"
    onSubmit={[Function]}
  >
    <h1>
      List your Item
    </h1>
    <input
      className="sign-on-input"
      label="name"
      name="name"
      onChange={[Function]}
      placeholder="Enter the item name:"
      required={true}
      type="text"
      value=""
    />
    <input
      className="sign-on-input"
      label="Price"
      name="price"
      onChange={[Function]}
      placeholder="Enter the price"
      required={true}
      type="number"
      value=""
    />
    <textarea
      className="sign-on-input add-items-description"
      label="description"
      name="description"
      onChange={[Function]}
      placeholder="Add your description"
      required={true}
      type="text"
      value=""
    />
    <p>
      Choose category
    </p>
    <div
      className="additems-categories"
      key="0"
    >
       
      electronics
      <input
        className=""
        label="category"
        name="category"
        onChange={[Function]}
        required={true}
        type="radio"
        value="electronics"
      />
       
    </div>
    <div
      className="additems-categories"
      key="1"
    >
       
      clothes
      <input
        className=""
        label="category"
        name="category"
        onChange={[Function]}
        required={true}
        type="radio"
        value="clothes"
      />
       
    </div>
    <div
      className="additems-categories"
      key="2"
    >
       
      shoes
      <input
        className=""
        label="category"
        name="category"
        onChange={[Function]}
        required={true}
        type="radio"
        value="shoes"
      />
       
    </div>
    <div
      className="additems-categories"
      key="3"
    >
       
      food
      <input
        className=""
        label="category"
        name="category"
        onChange={[Function]}
        required={true}
        type="radio"
        value="food"
      />
       
    </div>
    <div
      className="additems-categories"
      key="4"
    >
       
      general
      <input
        className=""
        label="category"
        name="category"
        onChange={[Function]}
        required={true}
        type="radio"
        value="general"
      />
       
    </div>
    <input
      className="sign-on-input"
      label="picture"
      name="picture"
      onChange={[Function]}
      placeholder="enter a imageUrl"
      required={true}
      type="url"
      value=""
    />
    <button
      onClick={[Function]}
      type="button"
    >
      VerifyImage
    </button>
    <a>
       Sign in to List Items
    </a>
  </form>
</div>
`;
